File: server/app/schemas/quality.py
"""Quality response models.

Schema note that shapes everything here: `quality_records` has two row shapes
(see `docs/database.md` section 5). A pass line has `defect_id IS NULL` and
carries the accepted units; each rejection line names one defect and carries the
units rejected for it. Summaries therefore aggregate across both shapes, while
the defect breakdown filters to `defect_id IS NOT NULL`.
"""

from __future__ import annotations

from datetime import date, datetime
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field

from app.models.enums import DefectSeverity


class QualityRecord(BaseModel):
    """One inspection line."""

    model_config = ConfigDict(from_attributes=True)

    id: UUID
    production_record_id: UUID
    inspected_at: datetime = Field(description="When the inspection was recorded (UTC).")

    machine_id: UUID
    machine_code: str
    component_id: UUID
    component_code: str
    component_name: str

    #: Null on the pass line, set on a rejection line.
    defect_id: UUID | None = None
    defect_code: str | None = None
    defect_name: str | None = None
    severity: DefectSeverity | None = None

    inspected_quantity: int = Field(ge=0)
    passed_quantity: int = Field(ge=0)
    rejected_quantity: int = Field(ge=0)
    first_pass_quantity: int = Field(
        ge=0, description="Units passing with no rework. Non-zero only on the pass line."
    )
    rework_quantity: int = Field(ge=0)

    is_rejection: bool = Field(description="True when this line records a defect.")


class QualitySummary(BaseModel):
    """Aggregate quality for a filtered period."""

    start_date: date
    end_date: date

    total_inspected: int = Field(ge=0, description="Units inspected across all lines.")
    total_passed: int = Field(ge=0)
    total_rejected: int = Field(ge=0)
    total_first_pass: int = Field(ge=0, description="Units that passed without rework.")
    total_rework: int = Field(ge=0)

    defect_rate_percentage: float = Field(ge=0, le=100, description="Rejected / inspected.")
    first_pass_yield_percentage: float = Field(
        ge=0, le=100, description="First-pass units / all units inspected."
    )
    quality_rate_percentage: float = Field(ge=0, le=100, description="Passed / inspected.")

    distinct_defect_types: int = Field(ge=0, description="Defect categories seen in the period.")
    record_count: int = Field(ge=0)
    has_data: bool


class DefectSummary(BaseModel):
    """One defect category's contribution, for the Pareto chart.

    `cumulative_percentage` is computed server-side. It depends on the position
    of a row within the sorted set, which is exactly the kind of thing that goes
    wrong when each chart component recalculates it from a differently sorted
    copy.
    """

    defect_id: UUID
    defect_code: str
    defect_name: str
    category: str
    default_severity: DefectSeverity

    rejected_quantity: int = Field(ge=0)
    occurrence_count: int = Field(ge=0, description="Inspection lines recording this defect.")
    share_percentage: float = Field(
        ge=0, le=100, description="This defect's share of all rejections."
    )
    cumulative_percentage: float = Field(
        ge=0, le=100, description="Running total of share, descending. Drives the Pareto line."
    )


class DefectTrendPoint(BaseModel):
    """One day on the defect trend chart."""

    bucket_date: date
    inspected_quantity: int = Field(ge=0)
    rejected_quantity: int = Field(ge=0)
    defect_rate_percentage: float = Field(ge=0, le=100)


class QualityByDimension(BaseModel):
    """Rejection totals grouped by machine, component or shift."""

    key_id: UUID
    key_code: str
    key_name: str
    inspected_quantity: int = Field(ge=0)
    rejected_quantity: int = Field(ge=0)
    defect_rate_percentage: float = Field(ge=0, le=100)

File: client/types/quality.ts
/**
 * Quality response models, mirroring `app/schemas/quality.py`.
 *
 * One schema fact shapes everything here: `quality_records` has two row shapes.
 * A pass line has `defect_id: null` and carries the accepted units; each
 * further line names one defect and carries its rejections. `is_rejection`
 * distinguishes them so a component never has to infer it from a null check.
 */
import type {
  DateRangeFilters,
  DefectSeverity,
  IsoDate,
  IsoDateTime,
  PaginationFilters,
  SortFilters,
  Uuid,
} from "@/types/common";

/** One inspection line. */
export interface QualityRecord {
  id: Uuid;
  production_record_id: Uuid;
  inspected_at: IsoDateTime;

  machine_id: Uuid;
  machine_code: string;
  component_id: Uuid;
  component_code: string;
  component_name: string;

  /** Null on the pass line, set on a rejection line. */
  defect_id: Uuid | null;
  defect_code: string | null;
  defect_name: string | null;
  severity: DefectSeverity | null;

  inspected_quantity: number;
  passed_quantity: number;
  rejected_quantity: number;
  /** Units passing with no rework. Non-zero only on the pass line. */
  first_pass_quantity: number;
  rework_quantity: number;

  /** True when this line records a defect. Derived server-side. */
  is_rejection: boolean;
}

/** Aggregate quality for a filtered period. */
export interface QualitySummary {
  start_date: IsoDate;
  end_date: IsoDate;

  total_inspected: number;
  total_passed: number;
  total_rejected: number;
  total_first_pass: number;
  total_rework: number;

  defect_rate_percentage: number;
  /**
   * First-pass units over *all* units inspected, not just those that passed.
   * Always at or below the quality rate; rework is what separates them.
   */
  first_pass_yield_percentage: number;
  quality_rate_percentage: number;

  distinct_defect_types: number;
  record_count: number;
  has_data: boolean;
}

/**
 * One defect category's contribution, for the Pareto chart.
 *
 * `cumulative_percentage` is computed server-side over the sorted set. It
 * depends on a row's position, which is exactly the thing that goes wrong when
 * each chart recalculates it from a differently sorted or truncated copy.
 */
export interface DefectSummary {
  defect_id: Uuid;
  defect_code: string;
  defect_name: string;
  category: string;
  default_severity: DefectSeverity;

  rejected_quantity: number;
  occurrence_count: number;
  share_percentage: number;
  cumulative_percentage: number;
}

/** One day on the defect trend chart. */
export interface DefectTrendPoint {
  bucket_date: IsoDate;
  inspected_quantity: number;
  rejected_quantity: number;
  defect_rate_percentage: number;
}

/** Rejection totals grouped by machine or component. */
export interface QualityByDimension {
  key_id: Uuid;
  key_code: string;
  key_name: string;
  inspected_quantity: number;
  rejected_quantity: number;
  defect_rate_percentage: number;
}

/** Filters accepted by the quality endpoints. */
export interface QualityFilters extends DateRangeFilters {
  machine_id?: Uuid;
  component_id?: Uuid;
  defect_id?: Uuid;
}

/** Quality list: filters plus pagination, sorting and the rejection switch. */
export interface QualityListFilters extends QualityFilters, PaginationFilters, SortFilters {
  /** Return only lines that recorded a defect. */
  rejections_only?: boolean;
}

/** Sort keys the quality list accepts. Mirrors the backend allow-list. */
export type QualitySortKey =
  "inspected_at" | "inspected" | "rejected" | "passed" | "machine" | "component";

/** Defect breakdown filters, which carry their own result limit. */
export interface DefectBreakdownFilters extends QualityFilters {
  limit?: number;
}

/** Dimensions quality totals can be grouped by. */
export type QualityDimension = "machine" | "component";
